(93) Display selected answers on confirmation

dgmstuart opened the pull request:
Previously we had a hardcoded confirmation page. Now use the Confirmation presenter object to get the answers and format them appropriately.

Of course this requires us to have actually stored the answers. In some cases this is simple (e.g. description has one field and no validations yet). In other cases it's fiddly (e.g. contact details has multiple fields and validations) so we delegate the saving of the answer to another object.

dgmstuart committed:
(feature) Display selected address on confirmation

- Pass one object to the view - most of what it displays will come from
  the selected answers.
- build this object with a hash. An alternative would be to pass a
  SelectedAnswerStore instance, or even just the session. Not sure where
  the knowledge about where the answers live should lie.
- stub property search to return postcode in other specs: this is that
  that those specs will fail correctly when validations _don't_ work: we
  want the failure to be that error isn't showing (because the
  confirmation page was loaded) rather than that the confirmation page
  blew up because it didn't have the data it needed (in this case:
  postcode)

dgmstuart committed:
(spec) Check all summary data vs hardcoded values

This will let us refactor in preparation for the next steps:
- Remove 'Type' For now - this isn't implemented

dgmstuart committed:
(spec) Failing: confirmation page (Varying data)

This is to force us to now test-drive out the actual functionality by
changing the data to be different from the hardcoded values.

dgmstuart committed:
(chore) Confirmation page summary text in locales

dgmstuart committed:
(feature) Confirmation includes contact details

Formatting telephone numbers is something we should do, but is
complicated and feels like a distraction at the moment.

dgmstuart committed:
(feature) Contact details can be persisted

A class with the responsibilities for 1. persisting the answers and 2.
knowing what data to put there. If we decide that this should be rolled
into the form objects then that should be fairly easy, but I like how
this way we get to keep a (tested) "valid?" method on the form and use
it in the "save" method here.

Maybe at some point the 'persist_answers' method gets extracted to its
own object, since the rest of the code is generic, but that feels like a
step too far at this stage.

dgmstuart committed:
(feature) Confirmation page shows contact details

These now come from the values saved by the user.

dgmstuart committed:
(feature) Raise errors on invalid callback time

In normal functioning we will prevent invalid callback times from
getting this far, but in case something goes wrong, I'd prefer to raise
an error than return nil.

dgmstuart committed:
(feature) Confirmation includes description

dgmstuart committed:
(refactor) Use a form object for description

This doesn't have any validations (valid? will always be true), but it
will be useful for this controller to be the same shape as all the
others.

dgmstuart committed:
(feature) repair description is stored in answers

I don't think this warrants a separate object like ContactDetailsSaver
at this point, but when this has validations (e.g. description length)
then we can create one at that point

leeky commented on spec/features/residents_can_see_confirmation_of_repair_request_spec.rb:
```diff
@@ -2,8 +2,38 @@
+    property = {
+      'property_reference' => '00000503',
+      'short_address' => 'Ross Court 30',
+      'postcode' => 'E5 8TE',
```
Is the API `list_properties` endpoint being changed to return the postcode in the results, because it doesn't at present..?

dgmstuart replied:
Yes: https://trello.com/c/VNIlJMI7/113-the-properties-propertyref-endpoint-should-return-postcode

I'm assuming this endpoint will ultimately return significantly more data - it was only this minimal as a first stab.

I considered merging the postcode into the answer hash ourselves, but we'd have to pass it from the radio buttons, which could get quite funky.

leeky commented on spec/features/residents_can_see_confirmation_of_repair_request_spec.rb:
```diff
@@ -28,12 +29,22 @@
+    expect(page).to have_content 'Phone: 01234 567 890'
     expect(page).to have_content 'Address: Ross Court 30, E5 8TE'
+
+    expect(page).to have_content 'Time: afternoon (12pm - 5pm)'
```
I think we should have a test for what happens when both callback times are checked, as I'm not sure what the output should look like.

dgmstuart replied:
Yup - that's covered later in the `Confirmation` presenter object: the approach I took was to test specifics in that object, and then have just one feature spec which tests the integration.

leeky replied:
:+1:

I did spot this after I posted my previous comment, but then forgot to go back and remove it!

leeky commented on app/views/confirmations/show.html.haml:
```diff
@@ -31,15 +31,13 @@
 
     %h3 Problem
-    %p Description: The bath won't drain
+    %p= t('confirmation.summary.description', description: "The bath won't drain")
```
:+1:

leeky commented on app/views/confirmations/show.html.haml:
```diff
@@ -31,15 +31,13 @@
 
     %h3 Callback time
-    %p Time: afternoon (12pm - 5pm)
+    %p= t('confirmation.summary.time', time: 'afternoon (12pm - 5pm)')
```
Is it worth renaming this to `callback_time`, as it might get confusing when we start dealing with `appointment_time` on the same page(?)

dgmstuart replied:
👍  Good shout - will do

leeky commented on app/controllers/describe_repair_controller.rb:
```diff
@@ -1,7 +1,17 @@
 class DescribeRepairController < ApplicationController
-  def index; end
+  def index
+    @form = DescriptionForm.new
```
Not sure if it matters, but I've been trying to keep the form and controller names in sync - so `DescribeRepairController` would reference `DescribeRepairForm`.

dgmstuart replied:
Sure, that makes sense 👍

leeky commented on spec/models/description_form_spec.rb:
```diff
@@ -0,0 +1,15 @@
+    end
+
+    it 'is valid with a description' do
+      expect(DescriptionForm.new(description: 'my taps are leaking').valid?).to eq true
```
How about splitting this into two lines to improve readability? i.e.

```ruby
form = DescriptionForm.new(description: 'my taps are leaking')
expect(form).valid?).to eq true
```

(NB: Because of RSpec's built-in matchers, you can also do `expect(form).to be_valid`)

dgmstuart commented:
Will pick up these tasks tomorrow

dgmstuart committed:
(refactor) Change 'time' to 'callback_time'

In future we'll have appointment times as well, so it's good to be clear
what we're talking about

dgmstuart committed:
(refactor) rename answer to 'describe repair'

The answer keys are generally related to the question page rather than
the answer.

dgmstuart committed:
(refactor) Rename DescriptionForm

All the others are named after the controller. Consistency hopefully
makes it easier to find things.

dgmstuart committed:
(refactor) DescribeRepairForm specs are tidier

dgmstuart commented:
@leeky I think I've addressed all the comments - ready for another look

dgmstuart committed:
(fix) Full name is shown in confirmation title

Missed this on the initial pass through

dgmstuart commented:
DON'T MERGE YET - spotted that the callback time isn't displayed properly in the body text

dgmstuart committed:
(feature) Callback time in confirmation body

It was previously only in the summary.

Changed the wording so that it makes sense in both contexts.

Using 'in the morning (8am - 12pm)' would be tricky, because of the all
day case: 'in the all day...' doesn't work, and 'Callback time: in the
morning ...' feels odd as well.

Decided against moving to locales at the moment: feels fiddly and we
have bigger fish to fry. Main thing for me right now is that the strings
aren't hardcoded into the view.
